File: Ruby/lib/Maybe.rb
class Maybe
    def initialize(value)
        @value = value
    end
    
    def value
        @value
    end
    
    # Functor
    def transform(&f)
        Maybe.new(@value.map { |x| f.call(x) })
    end

    def transform_(f)
        self.transform { |x| f.call(x) }
    end

    # Applicative
    def self.pure(x)
        Just(x)
    end

    def apply(&m)
        self.and_then { |f| m.call.transform { |x| f.call(x) } }
    end
    
    def *(m)
        self.apply { m }
    end
    
    # Monad
    def and_then(&f)
        @value.empty? ? Nothing : f.call(@value[0])
    end
    
    def skip(&m)
        self.and_then { |_| m.call }
    end
    
    def >>(m)
        self.skip { m }
    end

    # Alternative
    def self.empty
        Nothing
    end
    
    def or_else(&f)
        !@value.empty? ? self : f.call
    end

    def |(m)
        self.or_else { m }
    end
end

def Just(x)
    Maybe.new([x])
end

Nothing = Maybe.new([])

def Maybe_liftA2(f)
    lambda { |m1, m2| m1.transform { |x| f.call(x) } * m2 }
end

File: C#/Maybe.cs
namespace MonadParser
{
    public class Maybe
    {
        /// <summary>
        /// Wraps an existing value in an Maybe&lt;T&gt; instance.
        /// </summary>
        /// <param name="value">The value to be wrapped.</param>
        /// <returns>An optional containing the specified value.</returns>
        public static Maybe<T> Some<T>(T value) => new Maybe<T>(value, true);

        /// <summary>
        /// Creates an empty Maybe&lt;T&gt; instance.
        /// </summary>
        /// <returns>An empty optional.</returns>
#pragma warning disable 8604
        public static Maybe<T> None<T>() => new Maybe<T>(default(T), false);
#pragma warning restore 8604
    }

    public struct Maybe<T>
    {
        private readonly bool hasValue;
        private readonly T value;

        /// <summary>
        /// Checks if a value is present.
        /// </summary>
        public bool HasValue => hasValue;

        internal T Value => value;
        internal Maybe(T value, bool hasValue)
        {
            this.value = value;
            this.hasValue = hasValue;
        }

        // Alternative
        public static Maybe<T> Empty() => Maybe.None<T>();

        public Maybe<T> OrElse(Maybe<T> other) => hasValue ? this : other;

        public static Maybe<T> operator |(Maybe<T> m, Maybe<T> other) => m.OrElse(other);

        public Maybe<T> OrElseGet(Func<Maybe<T>> fm) => hasValue ? this : fm();

        public static Maybe<T> operator |(Maybe<T> m, Func<Maybe<T>> fm) => m.OrElseGet(fm);

        public TResult Match<TResult>(Func<T, TResult> some, Func<TResult> none) => hasValue ? some(value) : none();

        public Maybe<TResult> Map<TResult>(Func<T, TResult> f) => FlatMap(a => Maybe<TResult>.Pure(f(a)));

        public Maybe<TResult> FlatMap<TResult>(Func<T, Maybe<TResult>> mapping) => Match(
            some: mapping,
            none: Maybe.None<TResult>);

        // Applicative
        public static Maybe<T> Pure(T value) => Maybe.Some(value);

        public Maybe<TResult> Apply<TResult>(Maybe<Func<T, TResult>> f)
        {
            Maybe<T> self = this;
            return f.FlatMap(func => self.Map(func));
        }
    }
}
